### PR TITLE
Add support for more file storage APIs.

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
This change adds support for file storage operations on directories and
files; this includes creation, deletion and getting/setting of properties
and metadata.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
@mcardosos Please take a look.  :)

File: storage/file.go [hunk 1]
```diff
@@ -53,12 +101,83 @@ type ListSharesParameters struct {
...
+type httpVerb string
+type resourceType string
+
+const (
+	compNone          compType     = ""
+	compList          compType     = "list"
+	compMetadata      compType     = "metadata"
+	compProperties    compType     = "properties"
+	compRangeList     compType     = "rangelist"
+	httpGet           httpVerb     = "GET"
```
Comment (mcardosos) [inline storage/file.go]: Refactor! yesss! I bring [this](https://golang.org/pkg/net/http/#pkg-constants) to the table.

File: storage/file.go [hunk 2]
```diff
@@ -81,9 +192,94 @@ func (p ListSharesParameters) getParameters() url.Values {
...
+
+	var out DirsAndFilesListResponse
+	resp, err := f.listContent(path, q, nil)
+	if err != nil {
+		return out, err
+	}
+
+	defer resp.body.Close()
+	err = xmlUnmarshal(resp.body, &out)
+	return out, nil
```
Comment (mcardosos) [inline storage/file.go]: Aren't you returning the unmarshal error?

File: storage/file.go [hunk 3]
```diff
@@ -81,9 +192,94 @@ func (p ListSharesParameters) getParameters() url.Values {
...
+// See https://msdn.microsoft.com/en-us/library/azure/dn166980.aspx
+func (f FileServiceClient) ListDirsAndFiles(path string, params ListDirsAndFilesParameters) (DirsAndFilesListResponse, error) {
+	q := mergeParams(params.getParameters(), getURLInitValues(compList, resourceDirectory))
+
+	var out DirsAndFilesListResponse
+	resp, err := f.listContent(path, q, nil)
+	if err != nil {
+		return out, err
+	}
+
```
Comment (mcardosos) [inline storage/file.go]: Check for response code

File: storage/file.go [hunk 4]
```diff
@@ -81,9 +192,94 @@ func (p ListSharesParameters) getParameters() url.Values {
...
+		headers = make(map[string]string)
+		headers["Range"] = listRange.String()
+	}
+
+	var out FileRanges
+	resp, err := f.listContent(path, params, headers)
+	if err != nil {
+		return out, err
+	}
+
```
Comment (mcardosos) [inline storage/file.go]: I would move `defer resp.body.Close()` here.
What about the `Last-Modified` header?
The same as above, check for response code, and return the unmarshal error.

File: storage/file.go [hunk 5]
```diff
@@ -92,40 +288,171 @@ func pathForFileShare(name string) string {
...
+	resp, err := f.listContent("", q, nil)
 	if err != nil {
 		return out, err
 	}
-	defer resp.body.Close()
 
+	defer resp.body.Close()
 	err = xmlUnmarshal(resp.body, &out)
-	return out, err
+	return out, nil
```
Comment (mcardosos) [inline storage/file.go]: Return the error too

File: storage/file.go (hunk 2 above)
Comment (jhendrixMSFT) [inline storage/file.go]: Fixed.

File: storage/file.go (hunk 5 above)
Comment (jhendrixMSFT) [inline storage/file.go]: Fixed.

File: storage/file.go (hunk 3 above)
Comment (jhendrixMSFT) [inline storage/file.go]: I added the response code check to the listContent method so we only have to do it in one place (nice catch).

File: storage/file.go [hunk 6]
```diff
@@ -92,40 +292,176 @@ func pathForFileShare(name string) string {
...
+// See https://msdn.microsoft.com/en-us/library/azure/dn166993.aspx
+func (f FileServiceClient) CreateDirectory(path string) error {
+	return f.createResource(path, resourceDirectory, nil)
+}
+
+// CreateFile operation creates a new file or replaces an existing one.  Note that this
+// only initializes the file, call PutRange to add content.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dn194271.aspx
+func (f FileServiceClient) CreateFile(path string, maxSize uint64) error {
```
Comment (mcardosos) [inline storage/file.go]: I would add an extraHeaders parameter so other headers can be specified

File: storage/file.go [hunk 7]
```diff
@@ -92,40 +292,176 @@ func pathForFileShare(name string) string {
...
+}
+
+// CreateDirectory operation creates a new directory in the specified share. If a
+// directory with the same name already exists, the operation fails.
 //
-// See https://msdn.microsoft.com/en-us/library/azure/dn167008.aspx
-func (f FileServiceClient) CreateShare(name string) error {
-	resp, err := f.createShare(name)
+// See https://msdn.microsoft.com/en-us/library/azure/dn166993.aspx
+func (f FileServiceClient) CreateDirectory(path string) error {
```
Comment (mcardosos) [inline storage/file.go]: Add support for `x-ms-meta-name:value`

File: storage/file.go [hunk 8]
```diff
@@ -92,40 +292,176 @@ func pathForFileShare(name string) string {
...
+
+// modifies a range of bytes in the specified file
+func (f FileServiceClient) modifyRange(path string, bytes io.Reader, fileRange FileRange) error {
+	if err := f.checkForStorageEmulator(); err != nil {
+		return err
+	}
+	if fileRange.End < fileRange.Start {
+		return errors.New("the value for rangeEnd must be greater than or equal to rangeStart")
+	}
+	if fileRange.End-fileRange.Start > 4194304 {
```
Comment (mcardosos) [inline storage/file.go]: This constraint is just for `x-ms-write: update`

File: storage/file.go [hunk 9]
```diff
@@ -92,40 +292,176 @@ func pathForFileShare(name string) string {
...
+// See https://msdn.microsoft.com/en-us/library/azure/dn194276.aspx
+func (f FileServiceClient) ClearRange(path string, fileRange FileRange) error {
+	return f.modifyRange(path, nil, fileRange)
+}
+
+// PutRange writes a range of bytes to a file.  Note that the length of bytes must
+// match (rangeEnd - rangeStart) + 1 with a maximum size of 4MB.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dn194276.aspx
+func (f FileServiceClient) PutRange(path string, bytes io.Reader, fileRange FileRange) error {
```
Comment (mcardosos) [inline storage/file.go]: Will you add support for `Content-MD5` ?

File: storage/file.go [hunk 10]
```diff
@@ -92,40 +292,176 @@ func pathForFileShare(name string) string {
 // See https://msdn.microsoft.com/en-us/library/azure/dd179352.aspx
```
Comment (mcardosos) [inline storage/file.go]: This is the correct link `https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/list-shares`

Comment (jhendrixMSFT) [inline storage/file.go]: It appears that all of the links are redirecting to the new docs location.  Let's do a scrub of these in a separate PR.

File: storage/file.go (hunk 8 above)
Comment (jhendrixMSFT) [inline storage/file.go]: Fixed.

File: storage/file.go (hunk 7 above)
Comment (jhendrixMSFT) [inline storage/file.go]: CreateFile and CreateDirectory were modeled from CreateShare which doesn't allow one to specify metadata at the time of creation.  If we want to fix this for files and dirs I think we should also fix it for shares even though it's a breaking change.  What do you think?

Comment (mcardosos) [inline storage/file.go]: Oh D: maybe in a separate PR? all breaking changes together?

Comment (jhendrixMSFT) [inline storage/file.go]: Sure we could include this change as part of the v7 SDK update I'm working on.

File: storage/file.go (hunk 10 above)
Comment (mcardosos) [inline storage/file.go]: Agree

File: storage/file.go (hunk 7 above)
Comment (mcardosos) [inline storage/file.go]: It could be

File: storage/file.go [hunk 11]
```diff
@@ -92,40 +292,176 @@ func pathForFileShare(name string) string {
...
+			"Range": fileRange.String(),
+		}
+	}
+
+	resp, err := f.getResourceNoClose(path, compNone, resourceFile, http.MethodGet, extraHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = checkRespCode(resp.statusCode, []int{http.StatusOK, http.StatusPartialContent}); err != nil {
```
Comment (mcardosos) [inline storage/file.go]: PartialContent?

File: storage/file.go [hunk 12]
```diff
@@ -92,40 +292,176 @@ func pathForFileShare(name string) string {
...
+
 	defer resp.body.Close()
 	return checkRespCode(resp.statusCode, []int{http.StatusCreated})
 }
 
+// GetFile operation reads or downloads a file from the system, including its
+// metadata and properties.
+//
+// See https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/get-file
+func (f FileServiceClient) GetFile(path string, fileRange *FileRange) (*FileStream, error) {
```
Comment (mcardosos) [inline storage/file.go]: Just commenting this so we don't forget, this operation also supports md5.

**Comment by jhendrixMSFT** [comment]
@mcardosos Any other feedback before I merge?

File: storage/file.go [hunk 13]
```diff
@@ -53,12 +101,76 @@ type ListSharesParameters struct {
...
 // ShareHeaders contains various properties of a file and is an entry
 // in SetShareProperties
 type ShareHeaders struct {
 	Quota string `header:"x-ms-share-quota"`
 }
 
+type compType string
+type resourceType string
+
+const (
```
Comment (marstr) [inline storage/file.go]: What would you think of splitting this up into two const blocks to denote the two separate types represented here? Also, you may be able to bring this tighter to two enums by following the following pattern:
[printable-enums-in-go](http://stackoverflow.com/questions/23917916/printable-enums-in-go)

Thoughts?

Comment (jhendrixMSFT) [inline storage/file.go]: Good idea on splitting these into two blocks, I'll make that change.  I don't think switching to printable enums here is worth the added complexity (albeit not much) as for this case the string values are what's actually important (we place them into the request header).

**Comment by mcardosos** [comment]
LGTM
